packages/highlight-vdom/src/Highlight.tsx

```diff
@@ -0,0 +1,118 @@
+/** @jsx createElement */
+import {
+  cx,
+  ComponentChildren,
+  ComponentProps,
+  ElementType,
+  Renderer,
+} from '@algolia/ui-components-shared';
+
+type HighlightPartProps = {
+  children: ComponentChildren;
+  classNames: Partial<HighlightClassNames>;
+  highlightedTagName: ElementType;
+  nonHighlightedTagName: ElementType;
+  isHighlighted: boolean;
+};
+
+function createHighlightPartComponent({ createElement }: Renderer) {
+  return function HighlightPart({
+    classNames,
+    children,
+    highlightedTagName,
+    isHighlighted,
+    nonHighlightedTagName,
+  }: HighlightPartProps) {
+    const TagName = isHighlighted ? highlightedTagName : nonHighlightedTagName;
+
+    return (
+      <TagName
+        className={
+          isHighlighted ? classNames.highlighted : classNames.nonHighlighted
+        }
+      >
+        {children}
+      </TagName>
+    );
+  };
+}
+
+type HighlightedPart = {
+  isHighlighted: boolean;
+  value: string;
+};
+
+export type HighlightClassNames = {
+  /**
+   * Class names to apply to the root element
+   */
+  root: string;
+  /**
+   * Class names to apply to the highlighted parts
+   */
+  highlighted: string;
+  /**
+   * Class names to apply to the non-highlighted parts
+   */
+  nonHighlighted: string;
+  /**
+   * Class names to apply to the separator between highlighted parts
+   */
+  separator: string;
+};
+
+export type HighlightProps = ComponentProps<'span'> & {
+  classNames?: Partial<HighlightClassNames>;
+  highlightedTagName?: ElementType;
+  nonHighlightedTagName?: ElementType;
+  separator?: ComponentChildren;
+  parts: HighlightedPart[][];
+};
+
+export function createHighlightComponent({
+  createElement,
+  Fragment,
+}: Renderer) {
+  const HighlightPart = createHighlightPartComponent({
+    createElement,
+    Fragment,
+  });
+
+  return function Highlight({
+    parts,
+    highlightedTagName = 'mark',
+    nonHighlightedTagName = 'span',
+    separator = ', ',
+    className,
+    classNames = {},
+    ...props
+  }: HighlightProps) {
+    return (
+      <span {...props} className={cx(classNames.root, className) || undefined}>
```

Comment:
`cx()` should return a string?

```suggestion
      <span {...props} className={cx(classNames.root, className)}>
```

Reply:
It does return an empty string right now if `classNames.root` and `className` are not passed. Without the short-circuit, we end up with an empty `class` attribute.

Reply:
If having an empty `class` is a problem, let's implement this globally in `cx()` then?

Reply:
Yes I agree that `cx` should return `undefined` when there's nothing, I just wasn't sure to change it globally without discussing it first since we use it everywhere.

Reply:
is having an empty string for class even a problem?

Reply:
It's not a problem per se, but why have it if it's useless?

Reply:
Addressed in [de485fc](https://github.com/algolia/ui-components/pull/2/commits/de485fcc1c79e0934a7bf69701cd7b4d709fb0ce).